Add check for token at parse_bare_function_type

lucbf opened the pull request:
previously the compiler emmited:

> expecting ‘(’ but ‘identifier’ found

now it emmits:

> unexpected token 'identifier' - expected bare function

code example:
```
impl Foo for
fn my_function() {
//do something
}
```

lucbf committed:
Handle IDENTIFIER at parse_bare_function_type(

gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h: new error

Signed-off-by: Lúcio Boari Fleury <lucboari@gmail.com>

lucbf committed:
Previously this test emmited expected '('

gcc/testsuite/ChangeLog:

	* rust/compile/invalid_impl_for.rs: triggers check for identifier

Signed-off-by: Lúcio Boari Fleury <lucboari@gmail.com>

P-E-P commented:
> Not sure if I'm doing things right, but I had to delete my previous branch to rebase to the current gccrs

No this doesn't feel right. You don't have to delete anything to rebase your work on top of master. If you maintain a master branch on your fork then you can simply fetch it from the main repository and rebase it, either with your local git or online using the "sync fork" button.

lucbf commented:
Alpine build couldn't run properly because It couldn't download gettext-0.22.tar.gz

powerboat9 commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -9670,6 +9670,17 @@ Parser<ManagedTokenSource>::parse_bare_function_type (
     return nullptr;
 
+  auto t = lexer.peek_token ();
+  if (t->get_id () == IDENTIFIER)
```
You should probably replace this condition with `!= LEFT_PAREN`

lucbf committed:
Catch tokens different from IDENTIFIER

gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h: (parse_bare_function_type) broadens check

gcc/testsuite/ChangeLog:

	* rust/compile/invalid_impl_for.rs: broadens regex

Signed-off-by: Lúcio Boari Fleury <lucboari@gmail.com>

lucbf committed:
Improve regex for test

gcc/testsuite/ChangeLog:

	* rust/compile/invalid_impl_for.rs: nothing should change in principle

Signed-off-by: Lúcio Boari Fleury <lucboari@gmail.com>

P-E-P commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -9670,6 +9670,17 @@ Parser<ManagedTokenSource>::parse_bare_function_type (
+    }
+
   if (!skip_token (LEFT_PAREN))
     return nullptr;
```
This condition can be removed, we've already checked for a `LEFT_PAREN` and returned a nullptr, we only need to skip it.

lucbf committed:
Check for LEFT_PAREN is now redundant

gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h: (parse_bare_function_type) remove redundancy

Signed-off-by: Lúcio Boari Fleury <lucboari@gmail.com>

powerboat9 commented:
Looks like there are some merge conflicts with #4322, which split `rust-parse-impl.h` into separate files. You'll probably have to manually move your changes into the file that `parse_bare_function_type` was moved into.
